src/peft/tuners/oft/layer.py

```diff
@@ -282,100 +404,320 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+            scaled_rotated_weight = scaled_rotated_weight.to(previous_dtype)
+            if self.base_layer.bias is not None:
+                self.base_layer.bias = self.base_layer.bias.to(previous_dtype)
+            result = F.linear(input=x, weight=scaled_rotated_weight, bias=self.base_layer.bias)
 
-        # don't add bias here, because the bias will be added after OFT forward
-        return torch.matmul(input, delta_weight)
+        result = result.to(previous_dtype)
+        return result
 
     def __repr__(self) -> str:
         rep = super().__repr__()
         return "oft." + rep
 
 
-class Conv2d(OFTLayer):
+class Conv2d(nn.Module, OFTLayer):
     """OFT implemented in Conv2d layer"""
 
     def __init__(
         self,
         base_layer: nn.Module,
-        adapter_name: str = "default",
-        r: int = 0,
+        adapter_name: str,
+        r: int = 8,
+        oft_block_size: int = 0,
+        fan_in_fan_out: bool = False,  # Set this to True if the layer to replace stores weight like (fan_in, fan_out)
         module_dropout: float = 0.0,
-        init_weights: bool = True,
+        coft: bool = False,
+        eps: float = 6e-5,
+        block_share: bool = False,
+        init_weights: Union[bool, str] = True,
         **kwargs,
-    ):
-        super().__init__(base_layer)
+    ) -> None:
+        super().__init__()
+        OFTLayer.__init__(self, base_layer)
+        self.fan_in_fan_out = fan_in_fan_out
 
-        # Create adapter and set it active
         self._active_adapter = adapter_name
-        self.update_layer(adapter_name, r, module_dropout, init_weights, **kwargs)
 
-    def _get_delta_activations(
-        self, adapter_name: str, input: torch.Tensor, *args: Any, **kwargs: Any
-    ) -> torch.Tensor:
-        delta_weight = self.get_delta_weight(adapter_name)
+        # Create adapter and set it active
+        self.update_layer(
+            adapter_name, r, oft_block_size, module_dropout, coft, eps, block_share, init_weights
+        )
+
+    def update_layer(
+        self, adapter_name, r, oft_block_size, module_dropout, coft, eps, block_share, init_weights
+    ):
+        """
+        Update the conv2d layer with trainable OFT weights.
+        """
+        # Initialize the MultiplicativeDropoutLayer for module_dropout > 0.0.
+        if module_dropout > 0.0:
+            oft_dropout_layer = MultiplicativeDropoutLayer(p=module_dropout)
+        else:
+            oft_dropout_layer = nn.Identity()
+        self.oft_dropout.update(nn.ModuleDict({adapter_name: oft_dropout_layer}))
 
+        # layer information from the base layer
         base_layer = self.get_base_layer()
-        base_weight = base_layer.weight.data
-        delta_weight = delta_weight[: base_weight.shape[0], : base_weight.shape[0]]
+        conv_filter_dim = self.in_features * base_layer.kernel_size[0] * base_layer.kernel_size[0]
+
+        if r == 0 and oft_block_size != 0:
```

Comment:
Same comment as above, could this be done in the `OFTConfig` init?

Reply:
Yes. Updated.